presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionProcess.java

```diff
@@ -0,0 +1,265 @@
+                planFragmentCodec,
+                taskUpdateRequestCodec,
+                serverInfoCodec,
+                taskManagerConfig.getInfoRefreshMaxWait());
+        this.taskManagerConfig = requireNonNull(taskManagerConfig, "taskManagerConfig is null");
+        this.systemConfig = requireNonNull(systemConfig, "systemConfig is null");
+        this.nodeConfig = requireNonNull(nodeConfig, "nodeConfig is null");
+        this.connectorConfig = requireNonNull(connectorConfig, "connectorConfig is null");
+        this.nativeExecutionTask = new NativeExecutionTask(session, planFragment, sources, this.workerClient, tableWriteInfo, executor, updateScheduledExecutor, taskManagerConfig);
+        this.errorRetryScheduledExecutor = requireNonNull(errorRetryScheduledExecutor, "errorRetryScheduledExecutor is null");
+        this.errorTracker = nativeExecutionRequestErrorTracker("NativeExecution", uri, maxErrorDuration, errorRetryScheduledExecutor, "getting native process status");
+    }
+
+    /**
+     * Starts the external native execution process. The method will be blocked by connecting to the native process's /v1/info endpoint with backoff retries until timeout.
+     */
+    public void start()
+            throws ExecutionException, InterruptedException, IOException
+    {
+        String executablePath = getProcessWorkingPath(SystemSessionProperties.getNativeExecutionExecutablePath(session));
+        String configPath = Paths.get(executablePath, String.valueOf(port)).toAbsolutePath().toString();
+
+        populateConfigurationFiles(configPath);
+        ProcessBuilder processBuilder = new ProcessBuilder(executablePath, "--v", "1", "--etc_dir", configPath);
+        processBuilder.redirectOutput(ProcessBuilder.Redirect.INHERIT);
+        processBuilder.redirectError(ProcessBuilder.Redirect.INHERIT);
+        try {
+            log.info("Launching %s \nConfig path: %s\n", executablePath, configPath);
+            process = processBuilder.start();
+        }
+        catch (IOException e) {
+            throw new PrestoException(NATIVE_EXECUTION_PROCESS_LAUNCH_ERROR, format("Cannot start %s", processBuilder.command()), e);
+        }
+
+        getServerInfoWithRetry().get();
+    }
+
+    public NativeExecutionTask getTask()
+    {
+        return nativeExecutionTask;
+    }
+
+    @VisibleForTesting
+    public SettableFuture<ServerInfo> getServerInfoWithRetry()
+    {
+        SettableFuture<ServerInfo> future = SettableFuture.create();
+        doGetServerInfo(future);
+        return future;
+    }
+
+    public void close()
+    {
+        if (process.isAlive()) {
+            process.destroy();
+        }
+    }
+
+    private URI getBaseUriWithPort(URI baseUri, int port)
+    {
+        return uriBuilderFrom(baseUri)
+                .port(port)
+                .build();
+    }
+
+    private int getAvailableTcpPort()
+            throws IOException
+    {
+        ServerSocket socket = new ServerSocket(0);
```

Comment:
This method is bound to have race conditions, we are opening and closing the port, and other thread could re-open the same port. Not sure if my understanding is right.

Reply:
Once we close the socket with a given port, the TCP will be in the `TIME_WAIT` states for certain time (minutes IIRC), so during the `TIME_WAIT` period, that port won't be selected again by other processes/threads. Although this can not avoid the race condition totally, it can largely reduce the likelihood IMO. Without a central coordinated mechanism (say the coordinator assigns unique ports to different workers), this probably the only feasible way to choose an available port by each worker independently. Happy to hear your thought here.